### PR TITLE
unsigned, PositiveAutoFields

### LINKED ISSUE #8853: Implement unsigned integer/auto field
As part of #3705 we should make sure that our model fields represent what we have on -dev, stage and prod.

We should be able to steal some ideas from https://github.com/django/django/pull/8183/

### PR DESCRIPTION
**Pull request by eviljeff**
tests pass so must be okay!  fixes #8853

### CODE REVIEW

File: src/olympia/core/db/base.py [hunk 1]
```diff
@@ -0,0 +1,49 @@
+from django.db.backends.mysql.base import (
+    CursorWrapper as MySQLCursorWrapper,
+    DatabaseWrapper as MySQLDBWrapper,
+    DatabaseIntrospection as MySQLDBIntrospection,
+    DatabaseSchemaEditor as MySQLDBSchemeEditor)
+
+
+class CursorWrapper(MySQLCursorWrapper):
+    def execute(self, query, args=None):
+        return super(CursorWrapper, self).execute(query, args)
```
Comment (eviljeff) [inline src/olympia/core/db/base.py]: Actually, class isn't necessary - I was debugging using it to verify the correct sql was being generated.  I will drop it along with `DatabaseWrapper.create_cursor` and `DatabaseSchemaEditor.column_sql` methods.

**Comment by diox** [comment]
2 general comments:
-  We need a test or more proving that django does generate the proper sql at syncdb (well, initial migration) time. Otherwise we know it doesn't break anything but we don't know whether it actually does anything either :p
- Are we sure we've got unsigned primary keys on *all* our models right now? I doubt it.

**Comment by EnTeQuAk** [comment]
> We need a test or more proving that django does generate the proper sql at syncdb (well, initial migration) time. Otherwise we know it doesn't break anything but we don't know whether it actually does anything either :p


There is some scary looking code (but it's better than it looks) in `src/olympia/translations/test_fields.py::test_translated_field_supports_migration` which may be of help for ideas on how to test these fields. It's not the same though but a start.

File: src/olympia/core/db/base.py [hunk 2]
```diff
@@ -0,0 +1,49 @@
...
+        return field_type
+
+
+class DatabaseSchemaEditor(MySQLDBSchemeEditor):
+    def create_model(self, model):
+        for field in model._meta.local_fields:
+            # Autoincrement SQL
+            if field.get_internal_type() == "PositiveAutoField":
+                autoinc_sql = self.connection.ops.autoinc_sql(
+                    model._meta.db_table, field.column)
```
Comment (EnTeQuAk) [inline src/olympia/core/db/base.py]: I'm not sure if all this is actually needed, shouldn't we be able to do that via the field already?

```python
class UnsignedAutoField(models.AutoField):
    def db_type(self, connection):
        return 'integer UNSIGNED AUTO_INCREMENT'

    def rel_db_type(self, connection):
        return 'integer UNSIGNED'
```

(untested… something like that?)

Comment (eviljeff) [reply src/olympia/core/db/base.py]: It should be unnecessary but there's seemingly some edge case where you can indicate auto-incrementing later. I'm tempted to delete it.  

>
```python
class UnsignedAutoField(models.AutoField):
    def db_type(self, connection):
        return 'integer UNSIGNED AUTO_INCREMENT'

    def rel_db_type(self, connection):
        return 'integer UNSIGNED'
```

We don't need to do this normally - the code in `Field` that builds the sql for the column adds on the `AUTO_INCREMENT`, etc.

**Comment by eviljeff** [comment]
> Are we sure we've got unsigned primary keys on all our models right now? I doubt it.

The answer was No.  Not at all.  So I got a db structure from prod from autrilla and specifically added all the models that had an unsigned primary key (spoiler - there's no pattern!) rather than adding it to ModelBase.

As for testing - it's ... challenging.  Django doesn't seem to like you trying to do arbitrary "migrations" - it likes you to use it's complicated django migration setup - so getting a migration that doesn't actually exist means hacking stuff.   Normally I'd just go down until I got to the underlying code and test that, but we'd end up literally just testing the Field sql output from the Field I added and I'm not sure that's enough to "prove" it works. (or is it?)  The attempt here (still with random print statements) was based on the `sql_migrate` command but it doesn't work yet.

```python
class TestPositiveAutoField(TestCase):

    def get_sql_for_migration(self, migration):
        # Load up an executor to get all the migration data
        executor = MigrationExecutor(connection)
        # Show begin/end around output only for atomic migrations
        self.output_transaction = migration.atomic

        # Make a plan that represents just the requested migrations and show
        # SQL for it
        print executor.loader.graph.nodes
        plan = [(migration, False)]
        sql_statements = executor.collect_sql(plan)
        return '\n'.join(sql_statements)

    def test_migration(self):
        fields = {
            'id': PositiveAutoField(primary_key=True),
        }

        migration = type(str('Migration'), (migrations.Migration,), {
            'operations': [
                migrations.CreateModel(
                    name='MyModel', fields=tuple(fields.items()),
                    bases=(models.Model,)
                ),
            ],
        })
        print migration

        result = self.get_sql_for_migration(migration)
        print result

        # assert 'int(10) unsigned AUTOINCREMENT' in result
```

File: src/olympia/amo/fields.py [hunk 3]
```diff
@@ -10,6 +11,16 @@
 from olympia.amo.widgets import ColorWidget
 
 
+class PositiveAutoField(models.AutoField):
```
Comment (diox) [inline src/olympia/amo/fields.py]: Can you add a comment somewhere explaining that it needs `olympia.core.db.base` to work? It's not necessarily obvious when you look at it in isolation.

**Pull request merged**